gazebo_plugins/src/gazebo_ros_wheel_slip.cpp

```diff
@@ -256,6 +272,105 @@ void GazeboRosWheelSlip::Load(gazebo::physics::ModelPtr _model, sdf::ElementPtr
       }
     }
   }
+
+  double wheel_spin_tolerance = 1.0e-3;
+  if (!_sdf->HasElement("wheel_spin_tolerance")) {
+    RCLCPP_INFO(
+      impl_->ros_node_->get_logger(), "Missing <wheel_spin_tolerance>, defaults to %f",
+      wheel_spin_tolerance);
+  } else {
+    wheel_spin_tolerance = _sdf->GetElement("wheel_spin_tolerance")->Get<double>();
+  }
+
+  double publisher_update_rate = 100.0;
+  if (!_sdf->HasElement("publisher_update_rate")) {
+    RCLCPP_INFO(
+      impl_->ros_node_->get_logger(),
+      "Missing <publisher_update_rate>, defaults to %f", publisher_update_rate);
+  } else {
+    publisher_update_rate = _sdf->GetElement("publisher_update_rate")->Get<double>();
+  }
+
+  if (publisher_update_rate > 0.0) {
+    impl_->publisher_update_period_ = 1.0 / publisher_update_rate;
+  } else {
+    impl_->publisher_update_period_ = 0.0;
+  }
+
+  impl_->publisher_last_update_time_ = _model->GetWorld()->SimTime();
+
+  impl_->slip_publisher_ = impl_->ros_node_->create_publisher<gazebo_msgs::msg::WheelSlip>(
+    "wheel_slip", 10);
+
+  auto on_update_callback = [this, wheel_spin_tolerance](
+    const gazebo::common::UpdateInfo & info) {
+
+    #ifdef IGN_PROFILER_ENABLE
+      IGN_PROFILE("GazeboRosWheelSlip publisher");
+    #endif
+
+      gazebo::common::Time current_time = info.simTime;
+
+      // If the world is reset, for example
+      if (current_time < impl_->publisher_last_update_time_) {
+        RCLCPP_INFO(impl_->ros_node_->get_logger(), "Negative sim time difference detected.");
+        impl_->publisher_last_update_time_ = current_time;
+      }
+
+      // Check period
+      double seconds_since_last_update = (
+        current_time - impl_->publisher_last_update_time_).Double();
+
+      if (seconds_since_last_update < impl_->publisher_update_period_) {
+        return;
+      }
+
+    #ifdef IGN_PROFILER_ENABLE
+      IGN_PROFILE_BEGIN("fill ROS message");
+    #endif
+
+      // Populate message
+      auto slip_msg = gazebo_msgs::msg::WheelSlip();
+
+      std::map<std::string, ignition::math::Vector3d> slips;
+      this->GetSlips(slips);
+      for (const auto & wheel : slips) {
+        auto name = wheel.first;
+        auto slip = wheel.second;
+        auto spin_speed = slip.Z();
+        auto long_vel = slip.X() + slip.Z();
+        auto lat_vel = slip.Y();
+        double long_slip;
+        if (fabs(spin_speed) < wheel_spin_tolerance) {
+          long_slip = 0;
+        } else {
+          long_slip = (long_vel - spin_speed) / spin_speed;
```

Comment:
I think this doesn't follow the typical sign convention for longitudinal slip, which corresponds to positive slip during acceleration and negative slip during deceleration

see equation 1 in the following paper for example:

* http://www-cdr.stanford.edu/dynamic/WheelSlip/SlmillerGerdesACC.pdf

I noticed in the demo world that the published values are negative, when the vehicles are both driving against gravity, which is like accelerating so I would expect a positive slip value